library(tidyverse)
library(data.table)
library(treeWAS)

set.seed(100)

##########################  Load Data #################################

# panaroo everything
panaroo_all <- read.table("./data/raw/pv/gene_presence_absence.Rtab", header=TRUE) %>%
  select(-SAL_LA7644AA_AS.scaffold)


##### Scoary input #####

## Might be more useful to reduce. IGRs were collapsed when their distribution was the same across *all* hosts
# at least 1k (1/3rd) size reduction in features... probably worth doing so


### Bovine ###

# Collapsed 
scoary_coll_bovine <- read.csv("./data/raw/pv/scoary_panaroo_out/collapsed/Bovine.results.csv", header = TRUE, sep=",", stringsAsFactors = FALSE)
scoary_coll_bovine_sig <- scoary_coll_bovine %>% filter(Bonferroni_p < 0.05)
# 1026 bovine

# Uncollapsed
scoary_uncoll_bovine <- read.csv("./data/raw/pv/scoary_panaroo_out/uncollapsed/Bovine.results.csv", header = TRUE, sep=",", stringsAsFactors = FALSE)
scoary_uncoll_bovine_sig <- scoary_uncoll_bovine %>% filter(Bonferroni_p < 0.05)
# 1483  bovine, reduction by 400


### Poultry ###

# Collapsed
scoary_coll_poultry <- read.csv("./data/raw/pv/scoary_panaroo_out/collapsed/Poultry.results.csv", header = TRUE, sep=",", stringsAsFactors = FALSE)
scoary_coll_poultry_sig <- scoary_coll_poultry %>% filter(Bonferroni_p < 0.05)
# 1564 poultry collapsed

# Uncollapsed
scoary_uncoll_poultry <- read.csv("./data/raw/pv/scoary_panaroo_out/uncollapsed/Poultry.results.csv", header = TRUE, sep=",", stringsAsFactors = FALSE)
scoary_uncoll_poultry_sig <- scoary_uncoll_poultry %>% filter(Bonferroni_p < 0.05)
# 2706 poultry, reduction by 1000


### Swine ###

# Collapsed
scoary_coll_swine <- read.csv("./data/raw/pv/scoary_panaroo_out/collapsed/Swine.results.csv", header = TRUE, sep=",", stringsAsFactors = FALSE)
scoary_coll_swine_sig <- scoary_coll_swine %>% filter(Bonferroni_p < 0.05)
# 1126 swine collapsed

# Uncollapsed
scoary_uncoll_swine <- read.csv("./data/raw/pv/scoary_panaroo_out/uncollapsed/Swine.results.csv", header = TRUE, sep=",", stringsAsFactors = FALSE)
scoary_uncoll_swine_sig <- scoary_uncoll_swine %>% filter(Bonferroni_p < 0.05)
# 1879, reduction by 700

### Human ##

# Collapsed
scoary_coll_human <- read.csv("./data/raw/pv/scoary_panaroo_out/collapsed/Human.results.csv", header = TRUE, sep=",", stringsAsFactors = FALSE)
scoary_coll_human_sig <- scoary_coll_human %>% filter(Bonferroni_p < 0.05)
# 843 collapsed

# Uncollapsed
scoary_uncoll_human <- read.csv("./data/raw/pv/scoary_panaroo_out/uncollapsed/Human.results.csv", header = TRUE, sep=",", stringsAsFactors = FALSE)
scoary_uncoll_human_sig <- scoary_uncoll_human %>% filter(Bonferroni_p < 0.05)
# 1273, reduction by 300


### Metadata ###

# All metadata
metadata_all <- read.csv("./data/raw/metadata/metadata_quast_all_merged_filtered_discarded_modified.csv", header=TRUE) %>%
  select("Filename","Source.Host")

# Non-clonal metadata
metadata_nonclonal  <- read.csv("./data/raw/metadata/metadata_quast_all_merged_filtered_discarded_modified_noclonal.csv", header=TRUE) %>%
  select("Filename","Source.Host")

####################################################################### Uncollapsed #######################################################

#### Scoary Plots ####

# Bovine
p_b <- ggplot(scoary_uncoll_bovine, aes(Gene,-log10(Bonferroni_p ))) +
  geom_point(alpha = 0.50) +
  geom_hline(yintercept = -log10(0.05), color = "red", linetype = "dashed") +
  theme( axis.text.x=element_blank(),axis.ticks.x=element_blank()) +
  labs(x="IGRs", y=" -log10(Benferooni-corrected p-value)", title="Benferooni-Corrected P values of Bovine PV Dataset", 
       subtitle="Data generated by panaroo and scoary")

ggsave(filename="./data/out/round_2/plots/pv_bovine_bp.png",plot=p_b, device=png(), dpi=600,units="in", height=8,width = 8)
dev.off()

# Poultry
p_p <- ggplot(scoary_uncoll_poultry, aes(Gene,-log10(Bonferroni_p ))) +
  geom_point(alpha = 0.50) +
  geom_hline(yintercept = -log10(0.05), color = "red", linetype = "dashed") +
  theme( axis.text.x=element_blank(),axis.ticks.x=element_blank()) +
  labs(x="IGRs", y=" -log10(Benferooni-corrected p-value)", title="Benferooni-Corrected P values of Poultry PV Dataset", 
       subtitle="Data generated by panaroo and scoary")

ggsave(filename="./data/out/round_2/plots/pv_poultry_bp.png", plot=p_p, device=png(), dpi=600,units="in", height=8,width = 8)
dev.off()

# Swine
p_s <- ggplot(scoary_uncoll_swine, aes(Gene,-log10(Bonferroni_p ))) +
  geom_point(alpha = 0.50) +
  geom_hline(yintercept = -log10(0.05), color = "red", linetype = "dashed") +
  theme( axis.text.x=element_blank(),axis.ticks.x=element_blank()) +
  labs(x="IGRs", y=" -log10(Benferooni-corrected p-value)", title="Benferooni-Corrected P values of Swine PV Dataset", 
       subtitle="Data generated by panaroo and scoary")

ggsave(filename="./data/out/round_2/plots/pv_swine_bp.png", plot=p_s, device=png(), dpi=600,units="in", height=8,width = 8)
dev.off()

# Human
p_h <- ggplot(scoary_uncoll_human, aes(Gene,-log10(Bonferroni_p ))) +
  geom_point(alpha = 0.50) +
  geom_hline(yintercept = -log10(0.05), color = "red", linetype = "dashed") +
  theme( axis.text.x=element_blank(),axis.ticks.x=element_blank()) +
  labs(x="IGRs", y=" -log10(Benferooni-corrected p-value)", title="Benferooni-Corrected P values of Human PV Dataset", 
       subtitle="Data generated by panaroo and scoary")
ggsave(filename="./data/out/round_2/plots/pv_human_bp.png", plot=p_h, device=png(), dpi=600,units="in", height=8,width = 8)
dev.off()

# In an actual gwas these would be too many but since it's just a simple filtering step to choose what gets pplugged in RF, dont want to be too conservative

#### Modify intergenic region Data ####

## There is a lot  overlap between some though not all of thei pv
#intersect(scoary_uncoll_bovine_sig$Gene,scoary_uncoll_poultry_sig$Gene)
#intersect(scoary_uncoll_bovine_sig$Gene,scoary_uncoll_swine_sig$Gene)
#intersect(scoary_uncoll_bovine_sig$Gene,scoary_uncoll_human_sig$Gene)
#intersect(scoary_uncoll_poultry_sig$Gene,scoary_uncoll_swine_sig$Gene)
#intersect(scoary_uncoll_bovine_sig$Gene,scoary_uncoll_human_sig$Gene)
#intersect(scoary_uncoll_swine_sig$Gene,scoary_uncoll_human_sig$Gene)


# Combine scoary sig proteins (no duplicates)
sig_PV <-unique(c(scoary_uncoll_bovine_sig$Gene, 
                  scoary_uncoll_poultry_sig$Gene, 
                  scoary_uncoll_swine_sig$Gene,
                  scoary_uncoll_human_sig$Gene ))
# about 3k pvs

# Filter sig from piggy results
panaroo_sig <- panaroo_all %>% filter(Gene %in% sig_PV)

# Transpose
panaroo_sig_trans <- data.table::transpose(panaroo_sig)
colnames(panaroo_sig_trans) <- panaroo_sig$Gene # set IGRs as column names
panaroo_sig_trans$Assembly <- colnames(panaroo_sig) # get assemblies to column
panaroo_sig_trans = panaroo_sig_trans[-1,] # remove header row


###### Combine IGRs with metadata and save #######

# all
panaroo_sig_met_all <- inner_join(panaroo_sig_trans, metadata_all, by=c("Assembly" = "Filename")) %>%
  column_to_rownames(var="Assembly")
write.table(panaroo_sig_met_all, file="./data/out/round_2/input_data/pv_all.tsv", sep="\t")

# livestock
panaroo_sig_met_bps <- inner_join(panaroo_sig_trans, metadata_all, by=c("Assembly" = "Filename")) %>%
  column_to_rownames(var="Assembly") %>%
  filter(Source.Host != "Human")
write.table(panaroo_sig_met_bps, file="./data/out/round_2/input_data/pv_bps.tsv", sep="\t")

# non-clonal 
panaroo_sig_met_nonclonal <- inner_join(panaroo_sig_trans, metadata_nonclonal, by=c("Assembly" = "Filename")) %>%
  column_to_rownames(var="Assembly")
write.table(panaroo_sig_met_nonclonal, file="./data/out/round_2/input_data/pv_nonclonal.tsv", sep="\t")

# non-clonal & livestock
panaroo_sig_met_nonclonal_bps <- inner_join(panaroo_sig_trans, metadata_nonclonal, by=c("Assembly" = "Filename")) %>%
  column_to_rownames(var="Assembly") %>%
  filter(Source.Host != "Human")
write.table(panaroo_sig_met_nonclonal_bps, file="./data/out/round_2/input_data/pv_nonclonal_bps.tsv", sep="\t")


####################################################################### Collapsed #######################################################


#### Scoary Plots ####

# Bovine
p_b_c <- ggplot(scoary_coll_bovine, aes(Gene,-log10(Bonferroni_p ))) +
  geom_point(alpha = 0.50) +
  geom_hline(yintercept = -log10(0.05), color = "red", linetype = "dashed") +
  theme( axis.text.x=element_blank(),axis.ticks.x=element_blank()) +
  labs(x="IGRs", y=" -log10(Benferooni-corrected p-value)", title="Benferooni-Corrected P values of Bovine PV Dataset (Collapsed)", 
       subtitle="Data generated by panaroo and scoary")

ggsave(filename="./data/out/round_2/plots/pv_c_bovine_bp.png",plot=p_b_c, device=png(), dpi=600,units="in", height=8,width = 8)
dev.off()

# Poultry
p_p_c <- ggplot(scoary_coll_poultry, aes(Gene,-log10(Bonferroni_p ))) +
  geom_point(alpha = 0.50) +
  geom_hline(yintercept = -log10(0.05), color = "red", linetype = "dashed") +
  theme( axis.text.x=element_blank(),axis.ticks.x=element_blank()) +
  labs(x="IGRs", y=" -log10(Benferooni-corrected p-value)", title="Benferooni-Corrected P values of Poultry PV Dataset (Collapsed)", 
       subtitle="Data generated by panaroo and scoary")

ggsave(filename="./data/out/round_2/plots/pv_c_poultry_bp.png", plot=p_p_c, device=png(), dpi=600,units="in", height=8,width = 8)
dev.off()

# Swine
p_s_c <- ggplot(scoary_coll_swine, aes(Gene,-log10(Bonferroni_p ))) +
  geom_point(alpha = 0.50) +
  geom_hline(yintercept = -log10(0.05), color = "red", linetype = "dashed") +
  theme( axis.text.x=element_blank(),axis.ticks.x=element_blank()) +
  labs(x="IGRs", y=" -log10(Benferooni-corrected p-value)", title="Benferooni-Corrected P values of Swine PV Dataset (Collapsed)", 
       subtitle="Data generated by panaroo and scoary")

ggsave(filename="./data/out/round_2/plots/pv_c_swine_bp.png", plot=p_s_c, device=png(), dpi=600,units="in", height=8,width = 8)
dev.off()

# Human
p_h_c <- ggplot(scoary_coll_human, aes(Gene,-log10(Bonferroni_p ))) +
  geom_point(alpha = 0.50) +
  geom_hline(yintercept = -log10(0.05), color = "red", linetype = "dashed") +
  theme( axis.text.x=element_blank(),axis.ticks.x=element_blank()) +
  labs(x="IGRs", y=" -log10(Benferooni-corrected p-value)", title="Benferooni-Corrected P values of Human PV Dataset (Collapsed)", 
       subtitle="Data generated by panaroo and scoary")
ggsave(filename="./data/out/round_2/plots/pv_c_human_bp.png", plot=p_h_c, device=png(), dpi=600,units="in", height=8,width = 8)
dev.off()

# Combine scoary sig proteins (no duplicates)
sig_PV_c <-unique(c(scoary_coll_bovine_sig$Gene, 
                    scoary_coll_poultry_sig$Gene, 
                    scoary_coll_swine_sig$Gene,
                    scoary_coll_human_sig$Gene ))
# about 1.8k pvs


# Get first member in collapsed cluster as representative
sig_PV_c_rep <- as.data.frame(sig_PV_c) %>% 
  separate(sig_PV_c, into = c("V1", "Extra"), sep="--",extra = "merge") %>% 
  select(-Extra)

# Filter sig from piggy results
panaroo_sig_c <- panaroo_all %>% filter(Gene %in% sig_PV_c_rep$V1)

# Transpose
panaroo_sig_trans_c <- data.table::transpose(panaroo_sig_c)
colnames(panaroo_sig_trans_c) <- panaroo_sig_c$Gene # set IGRs as column names
panaroo_sig_trans_c$Assembly <- colnames(panaroo_sig_c) # get assemblies to column
panaroo_sig_trans_c = panaroo_sig_trans_c[-1,] # remove header row

###### Combine IGRs with metadata and save #######

# all
panaroo_sig_c_met_all <- inner_join(panaroo_sig_trans_c, metadata_all, by=c("Assembly" = "Filename")) %>%
  column_to_rownames(var="Assembly")
write.table(panaroo_sig_c_met_all, file="./data/out/round_2/input_data/pv_c_all.tsv", sep="\t")

# livestock
panaroo_sig_c_met_bps <- inner_join(panaroo_sig_trans_c, metadata_all, by=c("Assembly" = "Filename")) %>%
  column_to_rownames(var="Assembly") %>%
  filter(Source.Host != "Human")
write.table(panaroo_sig_c_met_bps, file="./data/out/round_2/input_data/pv_c_bps.tsv", sep="\t")

# non-clonal 
panaroo_sig_c_met_nonclonal <- inner_join(panaroo_sig_trans_c, metadata_nonclonal, by=c("Assembly" = "Filename")) %>%
  column_to_rownames(var="Assembly")
write.table(panaroo_sig_c_met_nonclonal, file="./data/out/round_2/input_data/pv_c_nonclonal.tsv", sep="\t")

# non-clonal & livestock
panaroo_sig_c_met_nonclonal_bps <- inner_join(panaroo_sig_trans_c, metadata_nonclonal, by=c("Assembly" = "Filename")) %>%
  column_to_rownames(var="Assembly") %>%
  filter(Source.Host != "Human")
write.table(panaroo_sig_c_met_nonclonal_bps, file="./data/out/round_2/input_data/pv_c_nonclonal_bps.tsv", sep="\t")

#######################################################################  TREEWAS ####################################################################### 

#Make genetic dataset

#A `matrix` containing binary genetic data (whether this encodes SNPs, gene presence/absence, etc. is up to you). 
#Individuals should be in the rows, and genetic variables in the columns. 
#Both rows and columns must be appropriately labelled with unique names. 
pv_gwas_genetic <- data.table::transpose(panaroo_all, keep.names = "File") %>%
  column_to_rownames(var="File")
colnames(pv_gwas_genetic) <- panaroo_all$Gene # set IGRs as column names
pv_gwas_genetic$Assembly <- colnames(panaroo_all) # get assemblies to column
pv_gwas_genetic = pv_gwas_genetic[-1,] # remove header row
pv_gwas_genetic <- as.matrix(sapply(pv_gwas_genetic, as.numeric)) # convert to matrix


#A phenotypic variable
#A `vector`  or `factor` containing either a binary or continuous variable encoding the phenotype for each individual. 
#Each element should have a name that corresponds to the row labels of the genetic data matrix (order does not matter). 
metadata_host_file <- metadata_all %>% 
  select(Source.Host, Filename) %>%
  filter(Filename %in% rownames(pv_gwas_genetic))

pv_gwas_metadata <- as.factor(metadata_host_file$Source.Host)
names(pv_gwas_metadata) <- metadata_host_file$Filename

#phylogenetic tree
#An object of class `phylo` (from the *ape* package), containing a phylogenetic tree. Tip labels are required and must correspond to both the row labels of the genetic data matrix and the names of the phenotypic variable. Any additional individuals, including the outgroup if not under analysis, should be removed prior to running *treeWAS*. The tree can be either rooted or unrooted.
pv_gwas_tree <- ape::read.tree(file = "../3.1gubbins/gubbins_snp_sl1344_out/core.full.1612602291.final_tree.tre")
culled_tree <- drop.tip(pv_gwas_tree, c("Reference", "SAL_LA7644AA_AS.scaffold"))

treewas_results <- treeWAS(snps = pv_gwas_genetic, phen = pv_gwas_metadata, tree = culled_tree)
