library(tidyverse)
library(data.table)
set.seed(100)

########### Load data #############

# Piggy Everything
piggy_all<- read.table("./data/raw/igr/IGR_presence_absence.Rtab", header=TRUE)%>%
  select(-SAL_LA7644AA_AS.scaffold)


##### Scoary input #####
scoary_coll_bovine_sig<- read.csv("./data/raw/igr/scoary_piggy_out/collapsed/Bovine.results.csv", header = TRUE, sep=",", stringsAsFactors = FALSE) %>% 
  filter(Bonferroni_p < 0.05)

### Poultry ###
# Collapsed
scoary_coll_poultry_sig <- read.csv("./data/raw/igr/scoary_piggy_out/collapsed/Poultry.results.csv", header = TRUE, sep=",", stringsAsFactors = FALSE) %>% 
  filter(Bonferroni_p < 0.05)

# Uncollapse

### Swine ###

# Collapsed
scoary_coll_swine <- read.csv("./data/raw/igr/scoary_piggy_out/collapsed/Swine.results.csv", header = TRUE, sep=",", stringsAsFactors = FALSE)
scoary_coll_swine_sig <- scoary_coll_swine %>% filter(Bonferroni_p < 0.05)
# 1030 swine collapsed

# Uncollapsed
scoary_uncoll_swine <- read.csv("./data/raw/igr/scoary_piggy_out/uncollapsed/Swine.results.csv", header = TRUE, sep=",", stringsAsFactors = FALSE)
scoary_uncoll_swine_sig <- scoary_uncoll_swine %>% filter(Bonferroni_p < 0.05)
# 1213 swine, reduction by 170


### Human ##

# Collapsed
scoary_coll_human <- read.csv("./data/raw/igr/scoary_piggy_out/collapsed/Human.results.csv", header = TRUE, sep=",", stringsAsFactors = FALSE)
scoary_coll_human_sig <- scoary_coll_human %>% filter(Bonferroni_p < 0.05)
# 730 human collapsed

# Uncollapsed
scoary_uncoll_human <- read.csv("./data/raw/igr/scoary_piggy_out/uncollapsed/Human.results.csv", header = TRUE, sep=",", stringsAsFactors = FALSE)
scoary_uncoll_human_sig <- scoary_uncoll_human %>% filter(Bonferroni_p < 0.05)
# 803 human, reduction by 50


### Metadata ###

# All metadata
metadata_all <- read.csv("./data/raw/metadata/metadata_quast_all_merged_filtered_discarded_modified.csv", header=TRUE) %>%
  select("Filename","Source.Host")

# Non-clonal metadata
metadata_nonclonal  <- read.csv("./data/raw/metadata/metadata_quast_all_merged_filtered_discarded_modified_noclonal.csv", header=TRUE) %>%
  select("Filename","Source.Host")

####################################################################### Uncollapsed #######################################################

#### Scoary Plots ####

# Bovine
p_b <- ggplot(scoary_uncoll_bovine, aes(Gene,-log10(Bonferroni_p ))) +
  geom_point(alpha = 0.50) +
  geom_hline(yintercept = -log10(0.05), color = "red", linetype = "dashed") +
  theme( axis.text.x=element_blank(),axis.ticks.x=element_blank()) +
  labs(x="IGRs", y=" -log10(Benferooni-corrected p-value)", title="Benferooni-Corrected P values of Bovine IGR Dataset", 
       subtitle="Data generated by piggy and scoary")

ggsave(filename="./data/out/round_2/plots/igr_bovine_bp.png",plot=p_b, device=png(), dpi=600,units="in", height=8,width = 8)
dev.off()

# Poultry
p_p <- ggplot(scoary_uncoll_poultry, aes(Gene,-log10(Bonferroni_p ))) +
  geom_point(alpha = 0.50) +
  geom_hline(yintercept = -log10(0.05), color = "red", linetype = "dashed") +
  theme( axis.text.x=element_blank(),axis.ticks.x=element_blank()) +
  labs(x="IGRs", y=" -log10(Benferooni-corrected p-value)", title="Benferooni-Corrected P values of Poultry IGR Dataset", 
       subtitle="Data generated by piggy and scoary")

ggsave(filename="./data/out/round_2/plots/igr_poultry_bp.png", plot=p_p, device=png(), dpi=600,units="in", height=8,width = 8)
dev.off()

# Swine
p_s <- ggplot(scoary_uncoll_swine, aes(Gene,-log10(Bonferroni_p ))) +
  geom_point(alpha = 0.50) +
  geom_hline(yintercept = -log10(0.05), color = "red", linetype = "dashed") +
  theme( axis.text.x=element_blank(),axis.ticks.x=element_blank()) +
  labs(x="IGRs", y=" -log10(Benferooni-corrected p-value)", title="Benferooni-Corrected P values of Swine IGR Dataset", 
       subtitle="Data generated by piggy and scoary")

ggsave(filename="./data/out/round_2/plots/igr_swine_bp.png", plot=p_s, device=png(), dpi=600,units="in", height=8,width = 8)
dev.off()

# Human
p_h <- ggplot(scoary_uncoll_human, aes(Gene,-log10(Bonferroni_p ))) +
  geom_point(alpha = 0.50) +
  geom_hline(yintercept = -log10(0.05), color = "red", linetype = "dashed") +
  theme( axis.text.x=element_blank(),axis.ticks.x=element_blank()) +
  labs(x="IGRs", y=" -log10(Benferooni-corrected p-value)", title="Benferooni-Corrected P values of Human IGR Dataset", 
       subtitle="Data generated by piggy and scoary")
ggsave(filename="./data/out/round_2/plots/igr_human_bp.png", plot=p_h, device=png(), dpi=600,units="in", height=8,width = 8)
dev.off()

# In an actual gwas these would be too many but since it's just a simple filtering step to choose what gets pplugged in RF, dont want to be too conservative

#### Modify intergenic region Data for output ####

## There is some overlap between some though not all of thei igr
#intersect(scoary_uncoll_bovine_sig$Gene,scoary_uncoll_poultry_sig$Gene)
#intersect(scoary_uncoll_bovine_sig$Gene,scoary_uncoll_swine_sig$Gene)
#intersect(scoary_uncoll_bovine_sig$Gene,scoary_uncoll_human_sig$Gene)
#intersect(scoary_uncoll_poultry_sig$Gene,scoary_uncoll_swine_sig$Gene)
#intersect(scoary_uncoll_bovine_sig$Gene,scoary_uncoll_human_sig$Gene)
#intersect(scoary_uncoll_swine_sig$Gene,scoary_uncoll_human_sig$Gene)

# Combine scoary sig proteins (no duplicates)
sig_IGRs <-unique(c(scoary_uncoll_bovine_sig$Gene, 
                    scoary_uncoll_poultry_sig$Gene, 
                    scoary_uncoll_swine_sig$Gene,
                    scoary_uncoll_human_sig$Gene ))
# about 2k IGRs


# Filter sig from piggy results
piggy_sig <- piggy_all %>% filter(Gene %in% sig_IGRs)

# Transpose
piggy_sig_trans <- data.table::transpose(piggy_sig)
colnames(piggy_sig_trans) <- piggy_sig$Gene # set IGRs as column names
piggy_sig_trans$Assembly <- colnames(piggy_sig) # get assemblies to column
piggy_sig_trans = piggy_sig_trans[-1,] # remove header row


###### Combine IGRs with metadata and save #######

# all
piggy_sig_met_all <- inner_join(piggy_sig_trans, metadata_all, by=c("Assembly" = "Filename"))  %>%
  column_to_rownames(var="Assembly") 
write.table(piggy_sig_met_all, file="./data/out/round_2/input_data/igr_all.tsv", sep="\t")

# livestock
piggy_sig_met_bps <- inner_join(piggy_sig_trans, metadata_all, by=c("Assembly" = "Filename"))  %>%
  column_to_rownames(var="Assembly") %>%
  filter(Source.Host != "Human")
write.table(piggy_sig_met_bps, file="./data/out/round_2/input_data/igr_bps.tsv", sep="\t")

# non-clonal
piggy_sig_met_nonclonal <- inner_join(piggy_sig_trans, metadata_nonclonal, by=c("Assembly" = "Filename"))  %>%
  column_to_rownames(var="Assembly")
write.table(piggy_sig_met_nonclonal, file="./data/out/round_2/input_data/igr_nonclonal.tsv", sep="\t")

# non-clonal & livestock
piggy_sig_met_nonclonal_bps <- inner_join(piggy_sig_trans, metadata_nonclonal, by=c("Assembly" = "Filename"))  %>%
  column_to_rownames(var="Assembly") %>%
  filter(Source.Host != "Human")
write.table(piggy_sig_met_nonclonal_bps, file="./data/out/round_2/input_data/igr_nonclonal_bps.tsv", sep="\t")



####################################################################### Collapsed #######################################################

#### Scoary Plots ####

# Bovine
p_b_c <- ggplot(scoary_coll_bovine, aes(Gene,-log10(Bonferroni_p ))) +
  geom_point(alpha = 0.50) +
  geom_hline(yintercept = -log10(0.05), color = "red", linetype = "dashed") +
  theme( axis.text.x=element_blank(),axis.ticks.x=element_blank()) +
  labs(x="IGRs", y=" -log10(Benferooni-corrected p-value)", title="Benferooni-Corrected P values of Bovine IGR Dataset (Collapsed)", 
       subtitle="Data generated by piggy and scoary")

ggsave(filename="./data/out/round_2/plots/igr_bovine_coll_bp.png",plot=p_b_c, device=png(), dpi=600,units="in", height=8,width = 8)
dev.off()

# Poultry
p_p_c <- ggplot(scoary_coll_poultry, aes(Gene,-log10(Bonferroni_p ))) +
  geom_point(alpha = 0.50) +
  geom_hline(yintercept = -log10(0.05), color = "red", linetype = "dashed") +
  theme( axis.text.x=element_blank(),axis.ticks.x=element_blank()) +
  labs(x="IGRs", y=" -log10(Benferooni-corrected p-value)", title="Benferooni-Corrected P values of Poultry IGR Dataset (Collapsed)", 
       subtitle="Data generated by piggy and scoary")

ggsave(filename="./data/out/round_2/plots/igr_poultry_coll_bp.png", plot=p_p_c, device=png(), dpi=600,units="in", height=8,width = 8)
dev.off()

# Swine
p_s_c <- ggplot(scoary_coll_swine, aes(Gene,-log10(Bonferroni_p ))) +
  geom_point(alpha = 0.50) +
  geom_hline(yintercept = -log10(0.05), color = "red", linetype = "dashed") +
  theme( axis.text.x=element_blank(),axis.ticks.x=element_blank()) +
  labs(x="IGRs", y=" -log10(Benferooni-corrected p-value)", title="Benferooni-Corrected P values of Swine IGR Dataset (Collapsed)", 
       subtitle="Data generated by piggy and scoary")

ggsave(filename="./data/out/round_2/plots/igr_swine_coll_bp.png", plot=p_s_c, device=png(), dpi=600,units="in", height=8,width = 8)
dev.off()

# Human
p_h_c <- ggplot(scoary_coll_human, aes(Gene,-log10(Bonferroni_p ))) +
  geom_point(alpha = 0.50) +
  geom_hline(yintercept = -log10(0.05), color = "red", linetype = "dashed") +
  theme( axis.text.x=element_blank(),axis.ticks.x=element_blank()) +
  labs(x="IGRs", y=" -log10(Benferooni-corrected p-value)", title="Benferooni-Corrected P values of Human IGR Dataset (Collapsed)", 
       subtitle="Data generated by piggy and scoary")
ggsave(filename="./data/out/round_2/plots/igr_human_coll_bp.png", plot=p_h_c, device=png(), dpi=600,units="in", height=8,width = 8)
dev.off()

#### Modify intergenic region Data for output ####

# Combine scoary sig proteins (no duplicates)
sig_IGRs_c <-unique(c(scoary_coll_bovine_sig$Gene, 
                      scoary_coll_poultry_sig$Gene, 
                      scoary_coll_swine_sig$Gene,
                      scoary_coll_human_sig$Gene ))

# about 1.7k IGRs

# Get first member in collapsed cluster as representative
sig_IGRs_c_rep <- as.data.frame(sig_IGRs_c) %>% 
  separate(sig_IGRs_c, into = c("V1", "Extra"), sep="--",extra = "merge") %>% 
  select(-Extra)

# Filter sig from piggy results
piggy_sig_c <- piggy_all %>% filter(Gene %in% sig_IGRs_c_rep$V1) 

# Transpose
piggy_sig_trans_c <- data.table::transpose(piggy_sig_c)
colnames(piggy_sig_trans_c) <- piggy_sig_c$Gene # set IGRs as column names
piggy_sig_trans_c$Assembly <- colnames(piggy_sig_c) # get assemblies to column
piggy_sig_trans_c = piggy_sig_trans_c[-1,] # remove header row


###### Combine IGRs with metadata and save #######

# all
piggy_sig_met_all_c <- inner_join(piggy_sig_trans_c, metadata_all, by=c("Assembly" = "Filename"))  %>%
  column_to_rownames(var="Assembly")
write.table(piggy_sig_met_all_c, file="./data/out/round_2/input_data/igr_c_all.tsv", sep="\t")

# livestock
piggy_sig_met_bps_c <- inner_join(piggy_sig_trans_c, metadata_all, by=c("Assembly" = "Filename"))  %>%
  column_to_rownames(var="Assembly") %>%
  filter(Source.Host != "Human")
write.table(piggy_sig_met_bps_c, file="./data/out/round_2/input_data/igr_c_bps.tsv", sep="\t")

# non-clonal
piggy_sig_met_nonclonal_c <- inner_join(piggy_sig_trans_c, metadata_nonclonal, by=c("Assembly" = "Filename"))  %>%
  column_to_rownames(var="Assembly")
write.table(piggy_sig_met_nonclonal_c, file="./data/out/round_2/input_data/igr_c_nonclonal.tsv", sep="\t")

# non-clonal & livestock
piggy_sig_met_nonclonal_bps_c <- inner_join(piggy_sig_trans_c, metadata_nonclonal, by=c("Assembly" = "Filename"))  %>%
  column_to_rownames(var="Assembly") %>%
  filter(Source.Host != "Human")
write.table(piggy_sig_met_nonclonal_bps_c, file="./data/out/round_2/input_data/igr_c_nonclonal_bps.tsv", sep="\t")


#######################################################################  TREEWAS ####################################################################### 

#Make genetic dataset

#A `matrix` containing binary genetic data (whether this encodes SNPs, gene presence/absence, etc. is up to you). 
#Individuals should be in the rows, and genetic variables in the columns. 
#Both rows and columns must be appropriately labelled with unique names. 


#A phenotypic variable
#A `vector`  or `factor` containing either a binary or continuous variable encoding the phenotype for each individual. 
#Each element should have a name that corresponds to the row labels of the genetic data matrix (order does not matter). 


#phylogenetic tree
#An object of class `phylo` (from the *ape* package), containing a phylogenetic tree. Tip labels are required and must correspond to both the row labels of the genetic data matrix and the names of the phenotypic variable. Any additional individuals, including the outgroup if not under analysis, should be removed prior to running *treeWAS*. The tree can be either rooted or unrooted.




